File: LC100/lc84.py
"""
Given n non-negative integers representing the histogram's bar height
 where the width of each bar is 1,
 find the area of largest rectangle in the histogram.

For example,
Given heights = [2,1,5,6,2,3],
return 10
"""


def largest(nums):
    """
    Calculate area for each bar as if it is the smallest one
     among the rectangle.
    """
    result = 0
    stack = [0]
    nums.append(0)

    for i in range(1, len(nums)):
        while stack and nums[i] < nums[stack[-1]]:
            height = nums[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            result = max(result, height * width)

        stack.append(i)

    return result


def test():
    nums_list = [
        [],
        [2, 1, 5, 6, 2, 3],
        [1, 1],
        [5, 1, 5],
        [1, 5, 1],
        [1, 2, 3, 4, 5],
        [4, 5, 6, 3, 6, 7, 8, 2]
    ]

    for nums in nums_list:
        print(largest(nums))


test()

File: LC100/lc84.js
function largest(nums) {
    nums.push(0);
    var stack = [0];
    var res = 0;
    var height, width;
    for (var i=1; i<nums.length; i++) {
        while (stack.length > 0 && nums[i] < nums[stack.slice(-1)[0]]) {
            height = nums[stack.pop()];
            if (stack.length === 0) {
                width = i;
            } else {
                width = i - 1 - stack.slice(-1)[0];
            }
            if (width * height > res) {
                res = width * height;
            }
        }
        stack.push(i);
    }
    return res;
}


function test() {
    var nums_list = [
        [],
        [2, 1, 5, 6, 2, 3],
        [1, 1],
        [5, 1, 5],
        [1, 5, 1],
        [1, 2, 3, 4, 5],
        [4, 5, 6, 3, 6, 7, 8, 2]
    ];
    for (var nums of nums_list) {
        console.log(largest(nums));
    }
}


test();
